cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs

```diff
@@ -5,89 +5,477 @@
+            DatabaseContext.ResidentEntities.Add(existingResidentNhsNumberMatch);
             DatabaseContext.SaveChanges();
-            var response = _classUnderTest.FindResident(findResidentCommand);
-            var createdRecord = DatabaseContext.ResidentEntities.Find(response);
-            var expectedRecord = new ResidentEntity
+
+            var nhsNumberMatchResidentId = DatabaseContext.ResidentEntities.FirstOrDefault(r => r.NhsNumber == matchingNhsNumber)?.Id;
+
+            // act
+            var duplicateResidentId = _classUnderTest.FindResident(searchParameters);
+            bool isResidentDuplicate = duplicateResidentId != null;
+
+            // assert
+            isResidentDuplicate.Should().BeTrue();
+            duplicateResidentId.Should().Be(nhsNumberMatchResidentId);
+        }
+
+        // If NHS number is empty or null or does not match &
+        // If Uprn is present, then:
+        // Resident is considered a duplicate, when First Name, Last Name & Uprn match.
+
+        [TestCase(null)]
+        [TestCase("")]
+        [TestCase(" ")]
+        [TestCase("nhs number")]
+        public void FindResidentReturnsAMatchWhenNoNhsNumberIsPresentAndWhenThereIsAResidentWithMatchingNonEmptyUprnAndFirstAndLastNamesForAllEmptyCasesOfNhsNumber(string testCaseNhsNumber)
+        {
+            // arrange
+            var matchingEmptyNhsNumber = testCaseNhsNumber;
+            var matchingFirstName = _faker.Random.Hash();
+            var matchingLastName = _faker.Random.Hash();
+            var matchingUprn = _faker.Random.Hash();
+
+            var searchParameters = new FindResident
             {
-                Id = existingResident.Id,
-                Uprn = "uprn",
-                FirstName = "FirstName",
-                LastName = "LastName"
+                NhsNumber = matchingEmptyNhsNumber,
+                Uprn = matchingUprn,
+                FirstName = matchingFirstName,
+                LastName = matchingLastName,
             };
 
-            createdRecord.Should().BeEquivalentTo(expectedRecord);
+            // empty nhs number match - making sure that doesn't get treated as a false positive
+            var existingResidentEmptyNhsNumberMatch = new ResidentEntity
+            {
+                NhsNumber = matchingEmptyNhsNumber,
+                Uprn = _faker.Random.Hash(),
+                FirstName = _faker.Random.Hash(),
+                LastName = _faker.Random.Hash()
+            };
+
+            var existingResidentUprnMatch = new ResidentEntity
+            {
+                NhsNumber = _faker.Random.Hash(),
+                Uprn = matchingUprn,
+                FirstName = matchingFirstName,
+                LastName = matchingLastName
+            };
+
+            // testing the no match path (adding this 1 line, so I wouldn't need to copy the whole test)
+            if (matchingEmptyNhsNumber == "nhs number") existingResidentEmptyNhsNumberMatch.NhsNumber = _faker.Random.Hash();
```

Comment:
This sounds like a separate test case.
Is this a valid way to do it ```[TestCase(_faker.Random.Hash())]```
Also, seems like ```matchingEmptyNhsNumber``` isn't always empty, is there something else we could name it, if you choose to keep it as a single test?

Reply:
Yeah the naming of this _("matchingEmptyNhsNumber")_ variable is a tricky business 😂 Will try to think of something better before merging.

I don't think that case deserves a separate test as it falls under the same edge case category as the others, which is that an NHS number de-duplication rule should find no match and fall through to the next one when it's either empty or simply does not match _(Essentially NHS rule should fall through to the next one on these 4 conditions - hence TestCases)_. An alternative is to copy paste this entire test for that one Test Case, which doesn't feel appropriate in terms of tests readability when this TestCase is the same category as the other ones.

Unfortunately, `[TestCase(_faker.Random.Hash())]` won't work, because the decorator attribute expects a concrete value. So dropping in a function call or an expression will give an error.